pkg/ipam-node/handlers/handlers.go

```diff
@@ -0,0 +1,136 @@
+
+// New create and initialize new instance of grpc Handlers
+func New(poolManager poolPkg.Manager, storeManager storePkg.Manager, getAllocFunc GetAllocatorFunc) *Handlers {
+	return &Handlers{
+		poolManager:  poolManager,
+		storeManager: storeManager,
+		getAllocFunc: getAllocFunc,
+	}
+}
+
+// Handlers contains implementation of the GRPC endpoints handlers for ipam-daemon
+type Handlers struct {
+	poolManager  poolPkg.Manager
+	storeManager storePkg.Manager
+	getAllocFunc GetAllocatorFunc
+	nodev1.UnimplementedIPAMBackendServiceServer
+}
+
+func (h *Handlers) openStore(ctx context.Context) (storePkg.Store, error) {
+	reqLog := logr.FromContextOrDiscard(ctx)
+	store, err := h.storeManager.Open(ctx)
+	if err != nil {
+		reqLog.Error(err, "failed to open store")
+		return nil, status.Errorf(codes.Internal, "failed to open store")
+	}
+	return store, nil
+}
+
+func (h *Handlers) closeStore(ctx context.Context, store storePkg.Store, err error) error {
+	reqLog := logr.FromContextOrDiscard(ctx)
+	if err == nil {
+		if err := store.Commit(); err != nil {
+			reqLog.Error(err, "failed to close store")
+			return status.Errorf(codes.Internal, "failed to close store")
+		}
+		reqLog.Info("all changes are committed to the store")
+		return nil
+	}
+	reqLog.Info("all store modifications are canceled")
+	store.Cancel()
+	return err
+}
+
+type paramsGetter interface {
+	GetParameters() *nodev1.IPAMParameters
+}
+
+func isRequiredError(field string) error {
+	return status.Errorf(codes.InvalidArgument, "%s is required parameter", field)
+}
+
+func isInvalidError(field string) error {
+	return status.Errorf(codes.InvalidArgument, "%s is invalid", field)
+}
+
+func validateReq(req paramsGetter) error {
+	params := req.GetParameters()
+	if params == nil {
+		return isRequiredError("parameters")
+	}
+	if len(params.Pools) == 0 || len(params.Pools) > 2 {
+		return isInvalidError("parameters.pools")
+	}
+	for i, p := range params.Pools {
+		if p == "" {
+			return isInvalidError(fmt.Sprintf("parameters.pools[%d]", i))
+		}
+	}
+	if params.CniContainerid == "" {
+		return isRequiredError("parameters.cni_containerid")
+	}
+	if params.CniIfname == "" {
+		return isRequiredError("parameters.cni_ifname")
+	}
+	if params.Metadata == nil {
+		return isRequiredError("parameters.metadata")
+	}
+	if params.Metadata.K8SPodName == "" {
+		return isRequiredError("parameters.metadata.k8s_pod_name")
+	}
+	if params.Metadata.K8SPodNamespace == "" {
+		return isRequiredError("parameters.metadata.k8s_pod_namespace")
+	}
+	return nil
+}
+
+func addFieldsToLogger(log logr.Logger, req paramsGetter) logr.Logger {
+	params := req.GetParameters()
+	if params == nil {
+		return log
+	}
+	return log.WithValues("pools", params.Pools,
+		"container_id", params.CniContainerid,
+		"interface_name", params.CniIfname,
+		"meta", params.Metadata.String(),
+	)
+}
+
+func checkReqISCanceled(ctx context.Context) error {
```

Comment:
nit: IS -> Is

Reply:
Why do we need this check ? to reduce the chance we have stale allocations ?

Reply:
> Why do we need this check ? to reduce the chance we have stale allocations ?

Yes. We are using global lock on the store level. Theoretically we can wait for a quite long before we will acquire the lock, so it is a good idea to check if we still need to do something.